### PR TITLE
8304837: Classfile API throws IOOBE for MethodParameters attribute without parameter names

### PR DESCRIPTION
**Pull request by SirYwell**
After merging master into https://github.com/openjdk/jdk/pull/9862, we encountered test failures (e.g., https://github.com/SirYwell/jdk/actions/runs/4500940829/jobs/7923018438#step:9:2541). The Classfile API tries to read from constant pool index 0 if a MethodParameters attribute has an entry without name.

The fix is simply using `readUtf8EntryOrNull` instead of `readUtf8Entry`. The related code already correctly handles nullability.

I didn't find an appropriate test class so I added a new one. Let me know if there's a better place or if the test can be improved somehow.

As I don't have a JBS account, someone needs to create a bug report there for me. Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304837](https://bugs.openjdk.org/browse/JDK-8304837): Classfile API throws IOOBE for MethodParameters attribute without parameter names

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - Committer)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13167/head:pull/13167` \
`$ git checkout pull/13167`

Update a local copy of the PR: \
`$ git checkout pull/13167` \
`$ git pull https://git.openjdk.org/jdk.git pull/13167/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13167`

View PR using the GUI difftool: \
`$ git pr show -t 13167`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13167.diff">https://git.openjdk.org/jdk/pull/13167.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13167#issuecomment-1481838703)

### CODE REVIEW

**Comment by liach** [comment]
Nice catch, issue opened at https://bugs.openjdk.org/browse/JDK-8304837 and marked as blocking compiler changes.

File: src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java:2]:
```suggestion
 * Copyright (c) 2022, 2023, Oracle and/or its affiliates. All rights reserved.
```

File: test/jdk/jdk/classfile/BoundAttributeTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import java.lang.constant.MethodTypeDesc;
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @summary Testing BoundAttributes
```
Comment (liach) [inline test/jdk/jdk/classfile/BoundAttributeTest.java:46]:
```suggestion
 * @issue 8304837
 * @summary Testing BoundAttributes
```
now that I've created an issue for you.

**Comment by TheShermanTanker** [comment]
Oops, looks like I was too late to create the issue haha

File: test/jdk/jdk/classfile/BoundAttributeTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import java.lang.constant.MethodTypeDesc;
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @issue 8304837
```
Comment (jonathan-gibbons) [inline test/jdk/jdk/classfile/BoundAttributeTest.java:46]: Should be `@bug`, not `@issue`. 

Did you run this test? Did it not complain at `@issue`?

File: test/jdk/jdk/classfile/BoundAttributeTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import java.lang.constant.ClassDesc;
+import java.lang.constant.ConstantDescs;
+import java.lang.constant.MethodTypeDesc;
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
```
Comment (jonathan-gibbons) [inline test/jdk/jdk/classfile/BoundAttributeTest.java:44]: Generally, the style for tests is to put the test description comment _before_ the imports.

File: test/jdk/jdk/classfile/BoundAttributeTest.java (hunk 3 above)
Comment (liach) [reply test/jdk/jdk/classfile/BoundAttributeTest.java:46]: sorry, i apparently suggested a wrong tag on github. my bad
```suggestion
 * @bug 8304837
```

**Comment by SirYwell** [comment]
Thanks for the review. I moved the comment and changed to `@bug`.

**Comment by TheShermanTanker** [comment]
Great work, feel free to @ me if you need a quick sponsor once you integrate it

**Comment by TheShermanTanker** [comment]
Keeping alive

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Comment by SirYwell** [comment]
Thank you for your review, @vicente-romero-oracle.

**Comment by TheShermanTanker** [comment]
Nice! :)

**Pull request closed without merging**